CONTRIBUTING.md

```diff
@@ -0,0 +1,221 @@
+- Fix a bug or an edge case.
+- Work on [![issues labeled good first issue](https://img.shields.io/github/issues/remeda/remeda/good%20first%20issue?style=flat-square)](https://github.com/remeda/remeda/issues?q=is%3Aopen+is%3Aissue+label%3A%22good+first+issue%22)
+- Work on [![issues labeled help wanted](https://img.shields.io/github/issues/remeda/remeda/help%20wanted?style=flat-square)](https://github.com/remeda/remeda/issues?q=is%3Aopen+is%3Aissue+label%3A%22help+wanted%22)
+
+## Getting started
+
+We use `npm` as the package manager. We recommend using [`nvm`](https://github.com/nvm-sh/nvm) to manage `Node.js` and `npm` versions. If you're not using `nvm`, the `.nvmrc` has the recommended Node version.
+
+Install dependencies:
+
+```bash
+git clone git@github.com:remeda/remeda.git
+cd remeda
+npm install
+```
+
+Run tests and watch for changes:
+
+```bash
+npm test # every test
+npm test -i src/file.test.ts # specific test
+```
+
+We have a pre-commit script that will format and lint the project, but if you want to run it manually:
+
+```bash
+npm run prettier # format
+npm run lint # lint
+```
+
+## Documentation
+
+Most of the content in the documentation is generated from the JSDoc comments. If you want to work on the design of the docs website or the migration guides, see [`docs/README.md`](https://github.com/remeda/remeda/blob/main/docs/README.md).
+
+## Guidelines
+
+### Adding a new function
+
+When adding a new function, remember to:
+
+- Add a JSDoc comment with a description, parameters, signature, an example, and tags.
+  - This will end up as the documentation on the website.
+- Add tests for runtime, typing, data-first, and data-last forms.
+  - See [Writing tests](#writing-tests) for guidelines on writing tests.
+- Add exports to `src/index.ts`.
+- Check for equivalent functions in [Lodash](https://lodash.com/docs/4.17.15) and [Ramda](https://ramdajs.com/docs/), and add them to `docs/src/content/mapping`.
+
+### Writing types
+
+We pay careful attention to types. Some tips for writing good types:
+
+- We use [`type-fest`](https://github.com/sindresorhus/type-fest) as our type utility library. We also have type utilities in [`internal/types`](src/internal/types.ts).
+- We use the [`Simplify`](https://github.com/sindresorhus/type-fest/blob/main/source/simplify.d.ts) type from `type-fest` for simplifying some output types.
+- We generally use the `IterableContainer` type from `internal/types` for input arrays. Similarly, we generally use the `object` type for input objects.
+
+<details>
+<summary>Examples:</summary>
+
+An example with an array:
+
+```ts
+const data = ["a", 1] as [string, number];
+
+// ❌ `T` can be too wide:
+function functionName<T>(data: ReadonlyArray<T>) {}
+
+functionName(data); // inside functionName, data[0] has type string | number
+
+// ✅ `T` is more specific:
+function functionName<T extends IterableContainer>(data: T) {}
+
+functionName(data); // inside functionName, data[0] has type string
+```
```

Comment:
The example is great, and I don't mind if we don't expand on this, but the main reason for using T as the whole array and not just the item type is that we maintain the _shape_ of the array in the output via mapping, so if the array is non-empty, we can return a non-empty output.

Reply:
i'll add a note for this. i was trying to think of more guidelines we can put in the writing types section but i haven't thought of any yet. i was thinking of reviewing past PR reviews and looking for patterns